File: src/server/core/acontext_core/schema/orm/session_event.py
from dataclasses import dataclass, field
from sqlalchemy import ForeignKey, Index, Column, String
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import JSONB, UUID
from typing import TYPE_CHECKING

from .base import ORM_BASE, CommonMixin
from ..utils import asUUID

if TYPE_CHECKING:
    from .session import Session
    from .project import Project


@ORM_BASE.mapped
@dataclass
class SessionEvent(CommonMixin):
    __tablename__ = "session_events"

    __table_args__ = (
        Index("idx_session_event_created", "session_id", "created_at"),
        Index("ix_session_event_project_id", "project_id"),
    )

    session_id: asUUID = field(
        metadata={
            "db": Column(
                UUID(as_uuid=True),
                ForeignKey("sessions.id", ondelete="CASCADE"),
                nullable=False,
            )
        }
    )

    project_id: asUUID = field(
        metadata={
            "db": Column(
                UUID(as_uuid=True),
                ForeignKey("projects.id", ondelete="CASCADE"),
                nullable=False,
            )
        }
    )

    type: str = field(
        metadata={
            "db": Column(String, nullable=False)
        }
    )

    data: dict = field(
        metadata={
            "db": Column(JSONB, nullable=False)
        }
    )

    # Relationships
    session: "Session" = field(
        init=False, metadata={"db": relationship("Session", back_populates="events")}
    )

    project: "Project" = field(
        init=False, metadata={"db": relationship("Project")}
    )

File: src/server/api/go/internal/modules/model/session_event.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

type SessionEvent struct {
	ID        uuid.UUID      `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	SessionID uuid.UUID      `gorm:"type:uuid;not null;index:idx_session_event_created,priority:1" json:"session_id"`
	ProjectID uuid.UUID      `gorm:"type:uuid;not null;index" json:"project_id"`
	Type      string         `gorm:"type:text;not null" json:"type"`
	Data      datatypes.JSON `gorm:"type:jsonb;not null" swaggertype:"object" json:"data"`

	CreatedAt time.Time `gorm:"autoCreateTime;not null;default:CURRENT_TIMESTAMP;index:idx_session_event_created,priority:2,sort:desc" json:"created_at"`
	UpdatedAt time.Time `gorm:"autoUpdateTime;not null;default:CURRENT_TIMESTAMP" json:"updated_at"`

	// SessionEvent <-> Session
	Session *Session `gorm:"foreignKey:SessionID;references:ID;constraint:OnDelete:CASCADE,OnUpdate:CASCADE;" json:"-"`

	// SessionEvent <-> Project
	Project *Project `gorm:"foreignKey:ProjectID;references:ID;constraint:OnDelete:CASCADE,OnUpdate:CASCADE;" json:"-"`
}

func (SessionEvent) TableName() string { return "session_events" }
